pyro/poutine/trace.py

```diff
@@ -20,105 +19,265 @@ def _warn_if_nan(name, value):
     # Note that -inf log_pdf is fine: it is merely a zero-probability event.
 
 
-class DiGraph(networkx.DiGraph):
-    node_dict_factory = collections.OrderedDict
+def topological_sort(trace):
+    """
+    Computes topological ordering of sites in the trace.
+    """
+    indegree_map = {node: 0 for node in trace.nodes}
+    for node_from, node_to in trace.iter_edges():
+        indegree_map[node_to] += 1
+    zero_indegree = [node for node in indegree_map if indegree_map[node] == 0]
+    while len(zero_indegree) > 0:
+        node = zero_indegree.pop()
+        for child in trace.edges[node]:
+            indegree_map[child] -= 1
+            if indegree_map[child] == 0:
+                zero_indegree.append(child)
+                del indegree_map[child]
+        yield node
+
+
+def map_trace(fn, nodes):
+    if isinstance(nodes, Trace):
+        # if we pass in a Trace, iterate over nodes
+        # to make things compositional
+        nodes = nodes.nodes
+    # return an OrderedDict to preserve input order
+    ret = collections.OrderedDict()
+    for name, node in nodes.items():
+        ret[name] = fn(node)
+    return ret
+
+
+def fold_trace(fn, initial_state, nodes):
+    if isinstance(nodes, Trace):
+        # if we pass in a Trace, iterate over nodes
+        # to make things compositional
+        nodes = nodes.nodes
+    state = initial_state
+    for name, node in nodes.items():
+        state = fn(state, node)
+    return state
+
+
+def filter_trace(fn, nodes):
+    if isinstance(nodes, Trace):
+        # if we pass in a Trace, iterate over nodes
+        # to make things compositional
+        nodes = nodes.nodes
+    # return an OrderedDict to preserve input order
+    ret = collections.OrderedDict()
+    for name, node in nodes.items():
+        if fn(node):
+            ret[name] = node
+    return ret
+
+
+def log_pdf(site):
+    try:
+        site_log_p = site["log_pdf"]
```

Comment:
`log_pdf()` would be a great name if we were using Haskell. Since we're using Python it would be better to name this `site_log_pdf()` or `get_log_pdf()` something narrower and verby.

Also, it's simpler to
```py
def log_pdf(site):
    try:
        return site["log_pdf"]
    except KeyError:
        ...compute and store...